module/zfs/dbuf.c

```diff
@@ -1593,12 +1594,18 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags,
 		ASSERT3S(db->db_state, ==, DB_NOFILL);
 
 		/*
-		 * Block cloning: If we have a pending block clone,
-		 * we don't want to read the underlying block, but the content
-		 * of the block being cloned, so we have the most recent data.
+		 * If we have a pending block clone, we don't want to read
+		 * the underlying block, but the content of the block being
+		 * cloned, pointed by the dirty record, so we have the most
+		 * recent data.  If there is no dirty record, then we hit a
+		 * race in a sync process when the dirty record is already
+		 * removed, while the dbuf is not yet destroyed.  Such case
+		 * is equivalent to uncached.
 		 */
 		dr = list_head(&db->db_dirty_records);
-		if (dr == NULL || !dr->dt.dl.dr_brtwrite) {
+		if (dr == NULL)
```

Comment:
Is it possible to set `db_state = DB_UNCACHED` in `dbuf_write_override_done` after testing that `db->state == DB_NOFILL` and no pending BRT clone in the next txg? That seems cleaner because then the dbuf has the same state it would have if the dbuf had been newly allocated by the reader hold.

Also, are there any lingering races/issues on the _write path_ during this racing hold window? e.g. write (or clone, or free) happens instead of read.

Reply:
> Is it possible to set `db_state = DB_UNCACHED` in `dbuf_write_override_done` after testing that `db->state == DB_NOFILL` and no pending BRT clone in the next txg?

That was my original thought too, as I have written in the description above, but if you look on dmu_buf_will_not_fill() and few other places, there are windows when dbuf lock is dropped between dbuf state change and dbuf_dirty() calls.  So I think there may be cases when DB_NOFILL not still set, but already set, and clearing it would be wrong.

> Also, are there any lingering races/issues on the _write path_ during this racing hold window? e.g. write (or clone, or free) happens instead of read.

I am not aware of any.  Though I have some bad feelings about correctness of db_state checks in dbuf_write(), since IMHO it should care only about dirty records from that specific transaction being synced, not a current dbuf state couple transactions later.  But I need more thinking to understand it better.

Reply:
I mean, there are.

That's one of the reasons for the panic in #11679 - on the write path, we set something we swear can't ever be in use by anyone else to NULL after writing before allocating a fresh buffer, and we're sometimes wrong and lose a race.

Reply:
> Though I have some bad feelings about correctness of db_state checks in dbuf_write()

After closer look I think #16057 should be cleaner there.

> I mean, there are.

@rincebrain On a quick look I am not sure how the issue and PR you mention are related to the issue of block cloning we hit, but I'll take a closer look.  If you just want to bring attention to the problem, there are cleaner ways.

Reply:
I wasn't trying to bring more attention to the issue, and I'm sorry if it came across that way; I just was reading the discussion and saw you voicing not knowing if there were races in `dbuf_write`, and I happened to know of an example, where, to the best of my understanding, you go down the `} else {` at the bottom of `dbuf_write` into `arc_write` which sets the abds to NULL on the way through, and then the panic I mentioned in #15538 happens if you go through `arc_buf_untransform_in_place` in the window between that and them being replaced afterward.

Reply:
Ah, okay. Thanks for explaining.

Here's a goto-free implementation for your consideration:

```
	blkptr_t bp, *bpp = NULL;

        ...

	if (db->db_state == DB_NOFILL) {
		dbuf_dirty_record_t *dr;

		/*
		 * Block cloning: If we have a pending block clone,
		 * we don't want to read the underlying block, but the content
		 * of the block being cloned, so we have the most recent data.
		 */
		dr = list_head(&db->db_dirty_records);
		if (dr != NULL) {
			if (!dr->dt.dl.dr_brtwrite) {
				err = EIO;
				goto early_unlock;
			}
			bp = dr->dt.dl.dr_overridden_by;
			bpp = &bp;
		}
	} else if (db->db_state != DB_CACHED) {
		err = EIO;
		goto early_unlock;
	}

	if (bpp == NULL && db->db_blkptr != NULL) {
		bp = *db->db_blkptr;
		bpp = &bp;
	}
```

Reply:
@rrevans Thanks. I've updated with some more compact version.